chips/sam4l/src/usbc/common_register.rs

```diff
@@ -0,0 +1,244 @@
+    pub fn read_word(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.addr) }
+    }
+}
+
+impl Reg<u32> {
+    #[inline]
+    pub fn set_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w | bit);
+    }
+
+    #[inline]
+    pub fn clr_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w & (!bit));
+    }
+}
+
+/// A write-only memory-mapped register, where any zero bits written have no effect
+pub struct RegW<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> RegW<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        RegW { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn write(self, val: T) {
+        unsafe { ::core::ptr::write_volatile(self.addr, T::to_word(val)); }
+    }
+}
+
+impl RegW<u32> {
+    #[inline]
+    pub fn set_bit(self, bit_index: u32) {
+        // For this kind of register, reads always return zero
+        // and zero bits have no effect, so we simply write the
+        // single bit requested.
+        let bit = 1 << bit_index;
+        self.write(bit);
+    }
+}
+
+/// A read-only memory-mapped register
+pub struct RegR<T> {
+    addr: *const u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> RegR<T> {
+    pub const fn new(addr: *const u32) -> Self {
+        RegR { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn read(self) -> T {
+        T::from_word(self.read_word())
+    }
+
+    #[inline]
+    pub fn read_word(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.addr) }
+    }
+}
+
+/// An array of memory-mapped registers
+pub struct Regs<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> Regs<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        Regs { addr: addr, phantom: PhantomData }
+    }
+
+    pub fn n(&self, index: u32) -> Reg<T> {
+        unsafe { Reg::new(self.addr.offset(index as isize)) }
+    }
+}
+
+/// An array of write-only memory-mapped registers
+pub struct RegsW<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> RegsW<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        RegsW { addr: addr, phantom: PhantomData }
+    }
+
+    pub fn n(&self, index: u32) -> RegW<T> {
+        unsafe { RegW::new(self.addr.offset(index as isize)) }
+    }
+}
+
+/// An array of read-only memory-mapped registers
+pub struct RegsR<T> {
+    addr: *const u32,
+    phantom: PhantomData<*const T>,
+}
+
+impl<T: FromWord + ToWord> RegsR<T> {
+    pub const fn new(addr: *const u32) -> Self {
+        RegsR { addr: addr, phantom: PhantomData }
+    }
+
+    pub fn n(&self, index: u32) -> RegR<T> {
+        unsafe { RegR::new(self.addr.offset(index as isize)) }
+    }
+}
+
+/// A bitfield of a memory-mapped register
+pub struct BitField<T> {
```

Comment:
Is the reason for this type basically that Rust doesn't have support for non-byte-aligned values in structs, and this lets you avoid doing binary arithmetic in that case? If so, I get it, but it seems _very_ dangerous to me from a type-safety perspective. _Perhaps_ if it only supported integral fields or something, but even then, we would have to think very hard about the implications...